amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -446,41 +455,128 @@ void runProcessLoop() {
+                    .stream().collect(Collectors
+                            .partitioningBy(streamIdentifier -> newStreamConfigMap.containsKey(streamIdentifier),
+                                    Collectors.toSet()));
+            final Set<StreamIdentifier> staleStreamIdsToBeDeleted = staleStreamIdDeletionDecisionMap.get(false).stream()
+                    .filter(streamIdentifier ->
+                            Duration.between(staleStreamDeletionMap.get(streamIdentifier), Instant.now()).toMillis()
+                                    >= getOldStreamDeferredDeletionPeriodMillis()).collect(Collectors.toSet());
+            streamsSynced.addAll(deleteMultiStreamLeases(staleStreamIdsToBeDeleted));
+
+            // Purge the active streams from stale streams list.
+            final Set<StreamIdentifier> staleStreamIdsToBeRevived = staleStreamIdDeletionDecisionMap.get(true);
+            removeActiveStreamsFromStaleStreamsList(staleStreamIdsToBeRevived);
+
             streamSyncWatch.reset().start();
         }
         return streamsSynced;
     }
 
+    @VisibleForTesting
+    long getOldStreamDeferredDeletionPeriodMillis() {
+        return OLD_STREAM_DEFERRED_DELETION_PERIOD_MILLIS;
+    }
+
     @VisibleForTesting
     boolean shouldSyncStreamsNow() {
         return isMultiStreamMode && (streamSyncWatch.elapsed(TimeUnit.MILLISECONDS) > NEW_STREAM_CHECK_INTERVAL_MILLIS);
     }
 
-    private void syncStreamsFromLeaseTableOnAppInit()
+    private void syncStreamsFromLeaseTableOnAppInit(List<MultiStreamLease> leases) {
+        final Set<StreamIdentifier> streamIdentifiers = leases.stream()
+                .map(lease -> StreamIdentifier.multiStreamInstance(lease.streamIdentifier()))
+                .collect(Collectors.toSet());
+        for (StreamIdentifier streamIdentifier : streamIdentifiers) {
+            if (!currentStreamConfigMap.containsKey(streamIdentifier)) {
+                currentStreamConfigMap.put(streamIdentifier, getDefaultStreamConfig(streamIdentifier));
+            }
+        }
+    }
+
+    private List<MultiStreamLease> fetchMultiStreamLeases()
             throws DependencyException, ProvisionedThroughputException, InvalidStateException {
-        if (!leasesSyncedOnAppInit && isMultiStreamMode) {
-            final Set<StreamIdentifier> streamIdentifiers = leaseCoordinator.leaseRefresher().listLeases().stream()
-                    .map(lease -> StreamIdentifier.multiStreamInstance(((MultiStreamLease) lease).streamIdentifier()))
-                    .collect(Collectors.toSet());
-            for (StreamIdentifier streamIdentifier : streamIdentifiers) {
-                if (!currentStreamConfigMap.containsKey(streamIdentifier)) {
-                    currentStreamConfigMap.put(streamIdentifier, getDefaultStreamConfig(streamIdentifier));
+        return (List<MultiStreamLease>) ((List) leaseCoordinator.leaseRefresher().listLeases());
```

Comment:
Why are we casting here multiple times?

Reply:
It is required to use "Untyped Casting" strategy for casting collections.